### PR TITLE
Fix some more low-hanging fruit in ManagedHandler

### PR DESCRIPTION
**Pull request by stephentoub**
- Setting the Host header in the request object incurs a variety of allocations.  We can skip that and just write the header directly to the output.
- Rather than writing strings char by char, if the whole string fits in the output buffer, we can just store the chars rather than doing an async write for each.
- Use cached header name strings, reason phrases, and some value strings, rather than always allocating a string for each
- Remove several of the intermediary async methods in the handler chain (though we should really try to remove them all... SendAsync is pretty much always going to yield, so this is adding measurable allocations on every call, even if it results in cleaner code)
- Use ContinueWith instead of async/await in a few places where ContinueWith results in less allocation
- Clean up the ManagedHandler class a bit, including ensuring arg validation is happening the same as in CurlHandler
- Fix several issues from the initial PR feedback

cc: @geoffkizer

### CODE REVIEW

File: src/System.Net.Http/src/System/Net/Http/Managed/HttpConnection.cs [hunk 1]
```diff
@@ -1204,5 +1212,32 @@ private void PutConnectionInPool()
 
             _pool.PutConnection(this);
         }
+
+        private struct ValueStringBuilder
+        {
+            public char[] Chars;
+            public int Offset;
```
Comment (justinvp) [inline src/System.Net.Http/src/System/Net/Http/Managed/HttpConnection.cs]: Nit: `Length` to align with `StringBuilder`? `Offset` makes me think of "start index" like `ArraySegment<T>.Offset`.

File: src/Common/src/System/Net/HttpKnownHeaderNames.TryGetHeaderName.cs [hunk 2]
```diff
@@ -8,6 +8,9 @@ namespace System.Net
 {
     internal static partial class HttpKnownHeaderNames
     {
+        private const string Gzip = "gzip";
```
Comment (danmosemsft) [inline src/Common/src/System/Net/HttpKnownHeaderNames.TryGetHeaderName.cs]: these strings already exist in the same project
```c#
    internal partial class CurlHandler : HttpMessageHandler
    {
        #region Constants

        private const char SpaceChar = ' ';
        private const int StatusCodeLength = 3;

        private const string UriSchemeHttp = "http";
        private const string UriSchemeHttps = "https";
        private const string EncodingNameGzip = "gzip";
        private const string EncodingNameDeflate = "deflate";
```

Maybe reuse them?

File: src/System.Net.Http/src/System/Net/Http/Managed/HttpConnectionKey.cs [hunk 3]
```diff
@@ -13,16 +13,16 @@ internal struct HttpConnectionKey : IEquatable<HttpConnectionKey>
         public HttpConnectionKey(Uri uri)
         {
             UsingSSL = 
-                uri.Scheme == "http" ? false :
-                uri.Scheme == "https" ? true :
+                uri.Scheme == UriScheme.Http ? false :
+                uri.Scheme == UriScheme.Https ? true :
                 throw new ArgumentException("Invalid Uri scheme", nameof(uri));
```
Comment (danmosemsft) [inline src/System.Net.Http/src/System/Net/Http/Managed/HttpConnectionKey.cs]: localize?

Comment (stephentoub) [inline src/System.Net.Http/src/System/Net/Http/Managed/HttpConnectionKey.cs]: Yes, there are a ton of strings that need to be moved to the .resx, consolidated, clean up, revised, etc.  I've not done so here and don't plan to in this PR.

File: src/Common/src/System/Net/HttpKnownHeaderNames.TryGetHeaderName.cs (hunk 2 above)
Comment (stephentoub) [inline src/Common/src/System/Net/HttpKnownHeaderNames.TryGetHeaderName.cs]: Yes, there is a ton more cleanup that can and should be done.  Note that in this case, I'm not actually increasing the number of duplicates... this duplicate already existed, I've just moved the file it was in.

File: src/System.Net.Http/src/System/Net/Http/Managed/HttpConnection.cs (hunk 1 above)
Comment (stephentoub) [inline src/System.Net.Http/src/System/Net/Http/Managed/HttpConnection.cs]: Well, it is a start index :)  It's the start of the next location available for writing, which just so happens to also be the current length because of the chars already written starting at 0.  But, eh, I can change it.

Comment (stephentoub) [inline src/System.Net.Http/src/System/Net/Http/Managed/HttpConnection.cs]: Fixed.

File: src/Common/src/System/Net/HttpKnownHeaderNames.TryGetHeaderName.cs [hunk 4]
```diff
@@ -47,6 +50,36 @@ public static unsafe bool TryGetHeaderName(IntPtr buffer, int length, out string
...
+            // If it's a known header value, use the known value instead of allocating a new string.
+
+            // Do a really quick reference equals check to see if name is the same object as
+            // HttpKnownHeaderNames.ContentEncoding, in which case the value is very likely to
+            // be either "gzip" or "deflate".
+            if (ReferenceEquals(name, ContentEncoding))
+            {
+                if (CharArrayHelpers.EqualsOrdinalAsciiIgnoreCase(Gzip, array, startIndex, length))
+                {
+                    return Gzip;
```
Comment (geoffkizer) [inline src/Common/src/System/Net/HttpKnownHeaderNames.TryGetHeaderName.cs]: If the server sends "GZIP" on the wire, then this will result in a header collection that contains "gzip".  I realize these are semantically equivalent, but do we care about preserving the literal bytes from the wire?

File: src/System.Net.Http/src/System/Net/Http/Managed/HttpConnection.cs [hunk 5]
```diff
@@ -575,14 +578,29 @@ public void Dispose()
...
+            }
+            char minorVersion = await ReadCharAsync(cancellationToken).ConfigureAwait(false);
+            if (!char.IsDigit(minorVersion))
+            {
+                throw new HttpRequestException("could not read response HTTP version");
+            }
+            response.Version =
+                (majorVersion == '1' && minorVersion == '1') ? HttpVersionInternal.Version11 :
+                (majorVersion == '1' && minorVersion == '0') ? HttpVersionInternal.Version10 :
+                (majorVersion == '2' && minorVersion == '0') ? HttpVersionInternal.Version20 :
```
Comment (geoffkizer) [inline src/System.Net.Http/src/System/Net/Http/Managed/HttpConnection.cs]: I don't think we should ever see 2.0 here.  2.0 will not be seen unless you do a connection upgrade.  That said, I guess it doesn't hurt to check for it.

File: src/System.Net.Http/src/System/Net/Http/Managed/HttpConnection.cs [hunk 6]
```diff
@@ -679,14 +704,10 @@ await ReadCharAsync(cancellationToken).ConfigureAwait(false) != '1')
                 headerName = headerName.TrimEnd();
 
                 // Add header to appropriate collection
-                // Don't ask me why this is the right API to call, but apparently it is
                 if (!response.Headers.TryAddWithoutValidation(headerName, headerValue))
                 {
-                    if (!responseContent.Headers.TryAddWithoutValidation(headerName, headerValue))
-                    {
-                        // Header name or value validation failed.
-                        throw new HttpRequestException($"invalid response header, {headerName}: {headerValue}");
```
Comment (geoffkizer) [inline src/System.Net.Http/src/System/Net/Http/Managed/HttpConnection.cs]: Why did you delete this?

Comment (stephentoub) [inline src/System.Net.Http/src/System/Net/Http/Managed/HttpConnection.cs]: Because none of the other handlers throw for this.

File: src/Common/src/System/Net/HttpKnownHeaderNames.TryGetHeaderName.cs (hunk 4 above)
Comment (stephentoub) [inline src/Common/src/System/Net/HttpKnownHeaderNames.TryGetHeaderName.cs]: We haven't. I guess in theory it could be important?  I can remove the ignore-case if we're concerned.

Comment (geoffkizer) [inline src/Common/src/System/Net/HttpKnownHeaderNames.TryGetHeaderName.cs]: I don't have a strong opinion, really.  It's certainly simpler internally to just normalize these, then we can do ref comparisons in a bunch of cases.  

I'd like to implement some optimizations to header handling in HttpClient proper, and being able to just always normalize the case of header names would be nice.

I'm trying to think of a case where the capitalization on the wire would be interesting to the user, and I'm failing.  Still, it might be a bit unexpected.

It looks like you stole this code from the WinHttpHandler, right?  Meaning we are already doing this normalization for WinHttpHandler?

File: src/System.Net.Http/src/System/Net/Http/Managed/HttpConnection.cs (hunk 6 above)
Comment (geoffkizer) [inline src/System.Net.Http/src/System/Net/Http/Managed/HttpConnection.cs]: OK.

File: src/Common/src/System/Net/HttpKnownHeaderNames.TryGetHeaderName.cs (hunk 4 above)
Comment (stephentoub) [inline src/Common/src/System/Net/HttpKnownHeaderNames.TryGetHeaderName.cs]: > Meaning we are already doing this normalization for WinHttpHandler?

Yes.

Comment (stephentoub) [inline src/Common/src/System/Net/HttpKnownHeaderNames.TryGetHeaderName.cs]: Let's go ahead with this for now. We can revisit it subsequently, and either undo it for both this and Windows, or potentially add more known values to the checks.

Comment (geoffkizer) [inline src/Common/src/System/Net/HttpKnownHeaderNames.TryGetHeaderName.cs]: Sounds good to me.
